test/test_cuda.py

```diff
@@ -901,6 +901,32 @@ def test_copy_device(self):
             self.assertEqual(z.get_device(), 0)
             self.assertIs(z.cuda(0), z)
 
+    @unittest.skipIf(not TEST_MULTIGPU, "only one GPU detected")
+    def test_copy_non_default_stream(self):
+        def _test_copy(self, x, y, output):
+            s0 = torch.cuda.Stream()
+            with torch.cuda.stream(s0):
+                torch.cuda._sleep(TestCuda.FIFTY_MIL_CYCLES)
+                y.copy_(x + 1, non_blocking=True)
+
+            s1 = torch.cuda.Stream()
+            with torch.cuda.stream(s1):
+                y.copy_(x, non_blocking=True)
+
+            s0.synchronize()
+            s1.synchronize()
+            self.assertEqual(y, output)
+
+        d0 = torch.device('cuda:0')
+        x0 = torch.zeros(5, 5, device=d0)
+
+        d1 = torch.device('cuda:1')
+        x1 = torch.zeros(5, 5, device=d1)
+        _test_copy(self, x0, x1, torch.zeros(5, 5, device=d1))
+
+        x2 = torch.zeros(5, 5, device=d0)
+        _test_copy(self, x0, x2, torch.ones(5, 5, device=d1))
```

Comment:
@colesbury 

This behavior is expected (as we do not sync if `src` and `dst` are the same), but looks weird to me. Should we do the sync using `dst` default stream regardless of whether `src` and `dst` are different?

Reply:
I don't understand. Which part is weird or unexpected here?

Reply:
These four lines, both are copying a zero tensor to `x0`, but the result differs, depending on whether the src and dst are on the same device. 
```python
x1 = torch.zeros(5, 5, device=d1)
_test_copy(self, x0, x1, torch.zeros(5, 5, device=d1))

x2 = torch.zeros(5, 5, device=d0)
_test_copy(self, x0, x2, torch.ones(5, 5, device=d1))
```

Reply:
To make the multi-device case like the single-device case you would have to set the streams for both d0 and d1 (and synchronize with the current stream instead of default stream). Something like:

```python
    def _test_copy(self, x, y, output):
        x_plus_one = x + 1

        s0 = torch.cuda.Stream()
        s1 = torch.cuda.Stream()
        s2 = torch.cuda.Stream(device=y.device)
        s3 = torch.cuda.Stream(device=y.device)

        with torch.cuda.stream(s2), torch.cuda.stream(s0):
                torch.cuda._sleep(50000000)
                y.copy_(x_plus_one)

        with torch.cuda.stream(s3), torch.cuda.stream(s1):
                y.copy_(x)

        s0.synchronize()
        s1.synchronize()
```